src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -519,21 +512,29 @@ public byte[] Http2AltSvcOriginUri
             Socket? socket = null;
             SslStream? sslStream = null;
             TransportContext? transportContext = null;
+            Http2Connection[]? lastHttp2Connections = _http2Connections;
 
             // Serialize creation attempt
             await _http2ConnectionCreateLock.WaitAsync(cancellationToken).ConfigureAwait(false);
             try
             {
-                if (_http2Connection != null)
+                Http2Connection[]? currentHttp2Connections = _http2Connections;
+                if (EnableMultipleHttp2Connections)
                 {
-                    // Someone beat us to it
-
-                    if (NetEventSource.Log.IsEnabled())
+                    if (_http2Connections != null)
                     {
-                        Trace("Using existing HTTP2 connection.");
+                        //Connection expiration will be checked here as well.
+                        Http2Connection? existingConnection = GetExistingHttp2Connection();
+                        if (existingConnection != null)
+                        {
+                            return (existingConnection, false, null);
+                        }
                     }
-
-                    return (_http2Connection, false, null);
+                }
+                else if (!ReferenceEquals(currentHttp2Connections, lastHttp2Connections) && currentHttp2Connections != null)
+                {
+                    // Someone beat us to it
+                    return (currentHttp2Connections[0], false, null);
```

Comment:
I don't think we need separate cases for multiple/single connections here, do we? Can't we just call GetExistingHttp2Connection and if it returns non-null, return that connection? Meaning, we don't need any of the logic here from line 521 to 537, just call GetExistingHttp2Connection.

Reply:
I left it because a simple reference comparison is much faster than doing expiration check in `GetExistingHttp2Connection`, but if the point here that all of these is negligible in comparison to a network call duration, then I agree it can be simplified to just one `GetExistingHttp2Connection` call.